flatgfa/src/packedseq.rs

```diff
@@ -0,0 +1,293 @@
+use std::ops::Index;
+use std::fmt;
+static A: u8 = 0;
+static C: u8 = 1;
+static G: u8 = 2;
+static T: u8 = 3;
+
+/// A compressed vector-like structure for storing nucleotide sequences
+///     - Two base pairs are stored per byte
+///
+/// data: a vector that stores a compressed encoding of this vec's sequence
+/// high_nibble_end: whether the final base pair in the sequence is stored at a 
+///                   high or low nibble 
+struct PackedVec {
+    data: Vec<u8>,
+    high_nibble_end: bool,
+}
+
+impl PackedVec {
+    fn new() -> Self {
+        PackedVec {
+            data: Vec::new(),
+            high_nibble_end: false,
+        }
+    }
+    pub fn push(&mut self, value: u8) {
+
+        if self.high_nibble_end {
+            self.data.push(value);
+            self.high_nibble_end = false;
+        } else {
+            let last_index = self.data.len() - 1;
+            self.data[last_index] = (value << 4) | self.data[last_index];
+            self.high_nibble_end = true;
+        }
+    }
+}
+
+// impl Index<usize> for PackedVec {
+//     type Output = u8;
+//     fn index(&self, index: usize) -> &Self::Output {
+//         if index % 2 == 1 {
+//             (&self.data[index as usize / 2] & 0b11110000u8) >> 4
+//         } else {
+//             &self.data[index as usize / 2] & 0b00001111u8
+//         }
+//     }
+// }
+
+impl fmt::Display for PackedVec {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "[")?;
+        let iter = PackedVecIterator {
+            data: &self.data,
+            cur_index: 0,
+            cur_high_nibble: false,
+            high_nibble_end: self.high_nibble_end,
+            last: false
+        };
+        let mut i = 0;
+        for item in iter {
+            if i == 0 {
+                write!(f, "{}", match item {0 => 'A', 1 => 'C', 2 => 'G', 3 => 'T', _ => ' '})?; 
```

Comment:
This kind of conversion to characters could be simplified by using a `Nucleotide` enum, as outlined above.